File: dart_sample/lib/demos/material/list_tile_controls_demo_page.dart
import 'package:flutter/material.dart';

class ListTileControlsDemoPage extends StatefulWidget {
  const ListTileControlsDemoPage({super.key});

  @override
  State<ListTileControlsDemoPage> createState() =>
      _ListTileControlsDemoPageState();
}

class _ListTileControlsDemoPageState extends State<ListTileControlsDemoPage> {
  bool _checkboxValue = false;
  bool? _tristateValue;
  bool _switchValue = true;
  bool _enabled = true;
  bool _adaptive = false;
  bool _compact = false;
  ListTileControlAffinity _affinity = ListTileControlAffinity.trailing;

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 10,
      children: <Widget>[
        const Text(
          'CheckboxListTile + SwitchListTile',
          style: TextStyle(fontSize: 20, color: Colors.black),
        ),
        const Text(
          'Whole-tile interaction, tristate cycle, affinity, density/alignment, selected styling, disabled state, and adaptive branches.',
          style: TextStyle(fontSize: 14, color: Colors.black54),
        ),
        Row(
          children: <Widget>[
            _buildControlButton(
              _compact ? 'Compact / top' : 'Standard / center',
              () => setState(() => _compact = !_compact),
              144,
              const Color(0xFFF3E5F5),
            ),
          ],
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildControlButton(
              _enabled ? 'Enabled' : 'Disabled',
              () => setState(() => _enabled = !_enabled),
              104,
              const Color(0xFFE9F0FF),
            ),
            _buildControlButton(
              _affinity == ListTileControlAffinity.leading
                  ? 'Leading'
                  : 'Trailing',
              () => setState(() {
                _affinity = _affinity == ListTileControlAffinity.leading
                    ? ListTileControlAffinity.trailing
                    : ListTileControlAffinity.leading;
              }),
              104,
              const Color(0xFFE9F7EF),
            ),
            _buildControlButton(
              _adaptive ? 'Adaptive' : 'Material',
              () => setState(() => _adaptive = !_adaptive),
              104,
              const Color(0xFFF8EFE2),
            ),
          ],
        ),
        Text(
          'checkbox=$_checkboxValue, tristate=$_tristateValue, switch=$_switchValue, affinity=${_affinity.name}, adaptive=$_adaptive',
          style: const TextStyle(fontSize: 12, color: Colors.blueGrey),
        ),
        Expanded(
          child: ColoredBox(
            color: const Color(0xFFF7F9FC),
            child: ListTileTheme(
              data: ListTileThemeData(
                controlAffinity: _affinity,
                visualDensity: _compact
                    ? VisualDensity.compact
                    : VisualDensity.standard,
                titleAlignment: _compact
                    ? ListTileTitleAlignment.top
                    : ListTileTitleAlignment.center,
              ),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.stretch,
                children: <Widget>[
                  _buildCheckboxTile(),
                  _buildTristateTile(),
                  _buildSwitchTile(),
                ],
              ),
            ),
          ),
        ),
      ],
    );
  }

  Widget _buildCheckboxTile() {
    final ValueChanged<bool?>? onChanged = _enabled
        ? (bool? value) => setState(() => _checkboxValue = value ?? false)
        : null;
    if (_adaptive) {
      return CheckboxListTile.adaptive(
        value: _checkboxValue,
        onChanged: onChanged,
        title: const Text('Wi-Fi discovery'),
        subtitle: const Text('Tap anywhere on the row to toggle.'),
        secondary: const Icon(Icons.info_outline),
        titleAlignment: _compact
            ? ListTileTitleAlignment.top
            : ListTileTitleAlignment.center,
        selected: _checkboxValue,
      );
    }
    return CheckboxListTile(
      value: _checkboxValue,
      onChanged: onChanged,
      title: const Text('Wi-Fi discovery'),
      subtitle: const Text('Tap anywhere on the row to toggle.'),
      secondary: const Icon(Icons.info_outline),
      titleAlignment: _compact
          ? ListTileTitleAlignment.top
          : ListTileTitleAlignment.center,
      selected: _checkboxValue,
      selectedTileColor: const Color(0xFFE8DEF8),
    );
  }

  Widget _buildTristateTile() {
    final ValueChanged<bool?>? onChanged = _enabled
        ? (bool? value) => setState(() => _tristateValue = value)
        : null;
    if (_adaptive) {
      return CheckboxListTile.adaptive(
        value: _tristateValue,
        onChanged: onChanged,
        tristate: true,
        title: const Text('Tristate selection'),
        subtitle: const Text('false → true → null → false'),
        secondary: const Icon(Icons.star_outline),
      );
    }
    return CheckboxListTile(
      value: _tristateValue,
      onChanged: onChanged,
      tristate: true,
      title: const Text('Tristate selection'),
      subtitle: const Text('false → true → null → false'),
      secondary: const Icon(Icons.star_outline),
    );
  }

  Widget _buildSwitchTile() {
    final ValueChanged<bool>? onChanged = _enabled
        ? (bool value) => setState(() => _switchValue = value)
        : null;
    if (_adaptive) {
      return SwitchListTile.adaptive(
        value: _switchValue,
        onChanged: onChanged,
        title: const Text('Background sync'),
        subtitle: const Text('The embedded switch remains draggable.'),
        secondary: const Icon(Icons.menu),
        selected: _switchValue,
      );
    }
    return SwitchListTile(
      value: _switchValue,
      onChanged: onChanged,
      title: const Text('Background sync'),
      subtitle: const Text('The embedded switch remains draggable.'),
      secondary: const Icon(Icons.menu),
      selected: _switchValue,
      selectedTileColor: const Color(0xFFE8DEF8),
    );
  }

  Widget _buildControlButton(
    String label,
    VoidCallback onPressed,
    double width,
    Color background,
  ) {
    return SizedBox(
      width: width,
      child: TextButton(
        onPressed: onPressed,
        style: TextButton.styleFrom(
          backgroundColor: background,
          foregroundColor: Colors.black,
          minimumSize: const Size(0, 36),
          padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 8),
          shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(8)),
        ),
        child: Text(label, style: const TextStyle(fontSize: 12)),
      ),
    );
  }
}

File: src/Sample/Plumix.Sample/Demos/Material/ListTileControlsDemoPage.cs
using System;
using Avalonia;
using Avalonia.Media;
using Plumix.Material;
using Plumix.Rendering;
using Plumix.Widgets;

namespace Plumix;

// Dart parity source (reference): dart_sample/lib/demos/material/list_tile_controls_demo_page.dart
public sealed class ListTileControlsDemoPage : StatefulWidget
{
    public override State CreateState() => new ListTileControlsDemoPageState();
}

internal sealed class ListTileControlsDemoPageState : State
{
    private bool _checkboxValue;
    private bool? _tristateValue;
    private bool _switchValue = true;
    private bool _enabled = true;
    private bool _adaptive;
    private bool _compact;
    private ListTileControlAffinity _affinity = ListTileControlAffinity.Trailing;

    public override Widget Build(BuildContext context)
    {
        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 10,
            children:
            [
                new Text("CheckboxListTile + SwitchListTile", fontSize: 20, color: Colors.Black),
                new Text(
                    "Whole-tile interaction, tristate cycle, affinity, density/alignment, selected styling, "
                    + "disabled state, and adaptive branches.",
                    fontSize: 14,
                    color: Color.Parse("#8A000000")),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildControlButton(
                            _enabled ? "Enabled" : "Disabled",
                            () => SetState(() => _enabled = !_enabled),
                            104,
                            Color.Parse("#FFE9F0FF")),
                        BuildControlButton(
                            _affinity == ListTileControlAffinity.Leading ? "Leading" : "Trailing",
                            () => SetState(() => _affinity = _affinity == ListTileControlAffinity.Leading
                                ? ListTileControlAffinity.Trailing
                                : ListTileControlAffinity.Leading),
                            104,
                            Color.Parse("#FFE9F7EF")),
                        BuildControlButton(
                            _adaptive ? "Adaptive" : "Material",
                            () => SetState(() => _adaptive = !_adaptive),
                            104,
                            Color.Parse("#FFF8EFE2")),
                    ]),
                new Row(
                    children:
                    [
                        BuildControlButton(
                            _compact ? "Compact / top" : "Standard / center",
                            () => SetState(() => _compact = !_compact),
                            144,
                            Color.Parse("#FFF3E5F5")),
                    ]),
                new Text(
                    $"checkbox={_checkboxValue.ToString().ToLowerInvariant()}, tristate={FormatNullable(_tristateValue)}, switch={_switchValue.ToString().ToLowerInvariant()}, affinity={_affinity.ToString().ToLowerInvariant()}, adaptive={_adaptive.ToString().ToLowerInvariant()}",
                    fontSize: 12,
                    color: Color.Parse("#FF607D8B")),
                new Expanded(
                    child: new Container(
                        color: Color.Parse("#FFF7F9FC"),
                        child: new ListTileTheme(
                            data: new ListTileThemeData(
                                ControlAffinity: _affinity,
                                VisualDensity: _compact ? VisualDensity.Compact : VisualDensity.Standard,
                                TitleAlignment: _compact
                                    ? ListTileTitleAlignment.Top
                                    : ListTileTitleAlignment.Center),
                            child: new Column(
                                crossAxisAlignment: CrossAxisAlignment.Stretch,
                                children:
                                [
                                    BuildCheckboxTile(),
                                    BuildTristateTile(),
                                    BuildSwitchTile(),
                                ])))),
            ]);
    }

    private Widget BuildCheckboxTile()
    {
        Action<bool?>? onChanged = _enabled
            ? value => SetState(() => _checkboxValue = value == true)
            : null;
        return _adaptive
            ? CheckboxListTile.Adaptive(
                value: _checkboxValue,
                onChanged: onChanged,
                title: new Text("Wi-Fi discovery"),
                subtitle: new Text("Tap anywhere on the row to toggle."),
                secondary: new Icon(Icons.InfoOutline),
                titleAlignment: _compact ? ListTileTitleAlignment.Top : ListTileTitleAlignment.Center,
                selected: _checkboxValue)
            : new CheckboxListTile(
                value: _checkboxValue,
                onChanged: onChanged,
                title: new Text("Wi-Fi discovery"),
                subtitle: new Text("Tap anywhere on the row to toggle."),
                secondary: new Icon(Icons.InfoOutline),
                titleAlignment: _compact ? ListTileTitleAlignment.Top : ListTileTitleAlignment.Center,
                selected: _checkboxValue,
                selectedTileColor: Color.Parse("#FFE8DEF8"));
    }

    private Widget BuildTristateTile()
    {
        Action<bool?>? onChanged = _enabled
            ? value => SetState(() => _tristateValue = value)
            : null;
        return _adaptive
            ? CheckboxListTile.Adaptive(
                value: _tristateValue,
                onChanged: onChanged,
                tristate: true,
                title: new Text("Tristate selection"),
                subtitle: new Text("false → true → null → false"),
                secondary: new Icon(Icons.StarOutline))
            : new CheckboxListTile(
                value: _tristateValue,
                onChanged: onChanged,
                tristate: true,
                title: new Text("Tristate selection"),
                subtitle: new Text("false → true → null → false"),
                secondary: new Icon(Icons.StarOutline));
    }

    private Widget BuildSwitchTile()
    {
        Action<bool>? onChanged = _enabled
            ? value => SetState(() => _switchValue = value)
            : null;
        return _adaptive
            ? SwitchListTile.Adaptive(
                value: _switchValue,
                onChanged: onChanged,
                title: new Text("Background sync"),
                subtitle: new Text("The embedded switch remains draggable."),
                secondary: new Icon(Icons.Menu),
                selected: _switchValue)
            : new SwitchListTile(
                value: _switchValue,
                onChanged: onChanged,
                title: new Text("Background sync"),
                subtitle: new Text("The embedded switch remains draggable."),
                secondary: new Icon(Icons.Menu),
                selected: _switchValue,
                selectedTileColor: Color.Parse("#FFE8DEF8"));
    }

    private static Widget BuildControlButton(string label, Action onPressed, double width, Color background)
    {
        return new SizedBox(
            width: width,
            child: new TextButton(
                onPressed: onPressed,
                child: new Text(label, fontSize: 12),
                style: TextButton.StyleFrom(
                    foregroundColor: Colors.Black,
                    backgroundColor: background,
                    padding: new Thickness(10, 8),
                    minimumSize: new Size(64, 36),
                    shape: new RoundedRectangleBorder(
                        borderRadius: BorderRadius.Circular(8)))));
    }

    private static string FormatNullable(bool? value)
    {
        return value.HasValue ? value.Value.ToString().ToLowerInvariant() : "null";
    }
}
